### DISCUSSION TITLE
Welcome to CeleScope Discussions!

### DISCUSSION THREAD

**Discussion by zhouyiqi91** [singleron-RD/CeleScope#27] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mehranfazel** [comment]
Hi, 

I use Telescope to analyze single cell RNA data generated by Singleron.
I want to extract the barcodes with following command but I receive this error. 
$ celescope rna  barcode  --fq1  /prj/ssc-ssrna-seq/row_data/example_1.fastq.gz  --fq2  /prj/ssc-ssrna-seq/row_data/Example_2.fastq.gz --outdir /prj/ssc-ssrna-seq/countTable --sample example  

-- celescope.tools.step.get_slot_key - WARNING - Assay not found in sample_summary.metrics
Traceback (most recent call last):
  File "/homes/mfazel/mambaforge/envs/celescope/bin/celescope", line 8, in <module>
    sys.exit(main())
...

i think its related with the --sample Attribute. 
can you please help me with this error? 

thanks!

**Reply by zhouyiqi91** [reply]
You need to run `sample` before `barcode`. The `barcode` step needs the `.metrics.json` file generated with the `sample` step.
![image](https://user-images.githubusercontent.com/30204951/208337261-f025def5-e609-4e69-90d7-ec3839f38a3a.png)
![image](https://user-images.githubusercontent.com/30204951/208337279-e1d989c7-1442-42f5-83bd-6ed8b771eadc.png)

**Comment by Acet221** [comment]
Hi,

I'm kind of new to bioinformatics and I wanted to try out CeleScope according to the user guide.
I have noticed that some packages in `conda_pkgs.txt` are only available for Linux/Mac, so I'm wondering if the whole workflow is designed to be done in Linux OS only, or if it might also be possible to run in Windows by replacing some packages for alternatives?

**Reply by zhouyiqi91** [reply]
CeleScope only runs on Linux systems. 
For example, CeleScope relys heavily on STAR which can not be run on Windows. https://github.com/alexdobin/STAR/issues/1428

**Comment by amarinderthind** [comment]
Hi, How to install it on public HPC system, with no conda command available?

**Reply by zhouyiqi91** [reply]
You can "fresh install" mamba without conda
https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html

**Reply by amarinderthind** [reply]
Thanks for that. Maenwhile, Wondering if you know something about this error that I encountered with standard installation (mamba;pip) on my desktop after pip installation command;   

``` File
"..SOme_PATH/miniconda3/envs/celescope/lib/python3.9/subprocess.py", line 524, in run
          raise CalledProcessError(retcode, process.args,
      subprocess.CalledProcessError: Command '['make', '-s', 'print-config']' returned non-zero exit status 2.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.
note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Reply by zhouyiqi91** [reply]
This error message is incomplete and does not show the specific error. Please attach the command and the full log.

**Reply by amarinderthind** [reply]
Full log  `pip install celescope 
Collecting celescope
  Using cached celescope-2.0.7-py3-none-any.whl.metadata (1.6 kB)
Collecting cutadapt==3.7 (from celescope)
  Using cached cutadapt-3.7-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (166 kB)
Collecting pysam>=0.18.0 (from celescope)
  Using cached pysam-0.22.0.tar.gz (4.6 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [42 lines of output]
      # pysam: cython is available - using cythonize if necessary
      # pysam: htslib mode is shared
      # pysam: HTSLIB_CONFIGURE_OPTIONS=None
      # pysam: (sysconfig) CC=gcc -pthread -B /home/asthind/miniforge3/envs/celescope/compiler_compat -Wl,--sysroot=/
      # pysam: (sysconfig) CFLAGS=-Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2   -isystem /home/asthind/miniforge3/envs/celescope/include -fPIC -O2   -isystem /home/asthind/miniforge3/envs/celescope/include  -fPIC
      # pysam: (sysconfig) LDFLAGS=-Wl,-rpath,/home/asthind/miniforge3/envs/celescope/lib -Wl,-rpath-link,/home/asthind/miniforge3/envs/celescope/lib -L/home/asthind/miniforge3/envs/celescope/lib -Wl,-rpath,/home/asthind/miniforge3/envs/celescope/lib -Wl,-rpath-link,/home/asthind/miniforge3/envs/celescope/lib -L/home/asthind/miniforge3/envs/celescope/lib
      checking for gcc... gcc -pthread -B /home/asthind/miniforge3/envs/celescope/compiler_compat -Wl,--sysroot=/
      checking whether the C compiler works... no
      configure: error: in `/tmp/pip-install-4ks3wrcb/pysam_9da6d68ab7ce426fbd0a81f760f6963a/htslib':
      configure: error: C compiler cannot create executables
      See `config.log' for more details
      checking for gcc... gcc -pthread -B /home/asthind/miniforge3/envs/celescope/compiler_compat -Wl,--sysroot=/
      checking whether the C compiler works... no
      configure: error: in `/tmp/pip-install-4ks3wrcb/pysam_9da6d68ab7ce426fbd0a81f760f6963a/htslib':
      configure: error: C compiler cannot create executables
      See `config.log' for more details
      Makefile:142: htscodecs.mk: No such file or directory
      /bin/sh: 1: ./hts_probe_cc.sh: not found
      make: *** [htscodecs.mk] Error 127
      # pysam: htslib configure options: None
      Traceback (most recent call last):
        File "/home/asthind/miniforge3/envs/celescope/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/home/asthind/miniforge3/envs/celescope/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/home/asthind/miniforge3/envs/celescope/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-xuzqycgb/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 325, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/tmp/pip-build-env-xuzqycgb/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 295, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-xuzqycgb/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 480, in run_setup
          super(_BuildMetaLegacyBackend, self).run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-xuzqycgb/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 442, in <module>
        File "<string>", line 82, in run_make_print_config
        File "/home/asthind/miniforge3/envs/celescope/lib/python3.9/subprocess.py", line 420, in check_output
          return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
        File "/home/asthind/miniforge3/envs/celescope/lib/python3.9/subprocess.py", line 524, in run
          raise CalledProcessError(retcode, process.args,
      subprocess.CalledProcessError: Command '['make', '-s', 'print-config']' returned non-zero exit status 2.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
`

**Reply by zhouyiqi91** [reply]
It may be related to the gcc you were using. https://github.com/velocyto-team/velocyto.py/issues/88

I think it can be fixed by installing pysam with mamba instead of pip
`mamba install -c bioconda pysam`
then
`pip install celescope`

**Reply by amarinderthind** [reply]
Thanks, seems `conda  install numpy scipy cython numba matplotlib scikit-learn h5py click
` worked.

**Comment by amarinderthind** [comment]
Hi, I was looking for information about index files, I got for each sample in context to scRNAseq. Do I need to provide this as an input somewhere? I could not find this info here https://github.com/singleron-RD/CeleScope/blob/master/doc/assay/multi_rna.md

**Reply by zhouyiqi91** [reply]
If your data is already demultiplexed (you have separate R1.fastq and R2.fastq files for each sample), then you don't really need it.
https://bioinformatics.stackexchange.com/questions/5178/what-is-the-index-fastq-file-sample-i-fastq-gz-generated-when-demultiplexing

**Reply by amarinderthind** [reply]
Great Got you. Thanks

**Comment by amarinderthind** [comment]
Where can I find the complete documentation for Celescope? I've received the final report, which includes two folders in the output—'raw' and 'filtered.' I'm unsure about which input to use for Seurat (raw or filtered). I couldn't find any documentation that explains the output details comprehensively, including the default filters applied.

Additionally, I'm curious about the tool or method Celescope uses for clustering cell types. When I use filtered data as input in Seurat, I observe different clusters than those generated by Celescope. Can you provide insights into this discrepancy?"

**Reply by zhouyiqi91** [reply]
The filter matrix is used for downstream analysis:
https://github.com/singleron-RD/CeleScope/blob/master/doc/assay/multi_rna.md#main-output
Since CeleScope uses Starsolo to generate matrices, you can refer to the Starsolo documentation to learn about the filtering algorithm:
https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md#emptydrop-like-filtering

CeleScope uses Scanpy to do the downstream analysis which may use different paramaters as what you use in Seurat. For example, the default resolution used in CeleScope(Scanpy) is 1.2.

**Reply by amarinderthind** [reply]
Thanks

**Comment by daniquebax** [comment]
I am running some DynaSCOPE samples and I wanted to get familiar with the celescope analysis. 
So I tried to run the test data, but I accounted some problems.

I first downloaded and installed the conda packages like on this site
`https://github.com/singleron-RD/CeleScope/blob/master/doc/user_guide.md`

Then I downloaded all the data and scripts like this
```
1. Get test data and scripts.
mkdir test_dir
cd test_dir
git clone https://github.com/singleron-RD/celescope_test_data.git
git clone https://github.com/singleron-RD/celescope_test_script.git
```

Then I generated the mouse genome like this:
```
Mus musculus
mkdir mmu_ensembl_99
cd mmu_ensembl_99

wget ftp://ftp.ensembl.org/pub/release-99/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.primary_assembly.fa.gz
wget ftp://ftp.ensembl.org/pub/release-99/gtf/mus_musculus/Mus_musculus.GRCm38.99.gtf.gz

gunzip Mus_musculus.GRCm38.dna.primary_assembly.fa.gz 
gunzip Mus_musculus.GRCm38.99.gtf.gz

conda activate celescope
celescope utils mkgtf Mus_musculus.GRCm38.99.gtf Mus_musculus.GRCm38.99.filtered.gtf

celescope rna mkref \
--genome_name Mus_musculus_ensembl_99_filtered \
--fasta Mus_musculus.GRCm38.dna.primary_assembly.fa \
--gtf Mus_musculus.GRCm38.99.filtered.gtf \
--mt_gene_list mt_gene_list.txt
```

Then I modified arguments in the dynaseq script to my mouse genomeDir path like this

```
multi_dynaseq \
    --mapfile ./case.mapfile \
    --genomeDir /scratch/dbax/CeleScope/mmu_ensembl_99 \
    --mod shell

```
Then I ran the test like this

```
cd /scratch/dbax/CeleScope/test_dir/celescope_test_script
(celescope)dbax@cn106:/scratch/dbax/CeleScope/test_dir/celescope_test_script$ python -m pytest -s test_multi.py --assays dynaseq
```

Sadly this is the point where I get an error message. I looked at the log file (attached 1) and it seems to be a problem that it cannot find the mt_gene_list and gtf files. However, when I go to my genomeDIR I do think that I see these files (attached 2)
[dynaseq.log](https://github.com/singleron-RD/CeleScope/files/14433582/dynaseq.log)
![dynaseq](https://github.com/singleron-RD/CeleScope/assets/127849963/06604e2f-5064-416f-b9fe-867b8e412bc0)

Do you know how I can fix this problem? I have limited bioinformatics experience, so it is difficult for me to logically describe my problem to you. I hope you understand my question.

**Reply by zhouyiqi91** [reply]
```
  File "/vol/mbconda/dbax/celescope/lib/python3.9/site-packages/celescope/tools/analysis_wrapper.py", line 64, in __init__
    self.mt_gene_list = Mkref_rna.get_config(args.genomeDir)['files']['mt_gene_list']
  File "/vol/mbconda/dbax/celescope/lib/python3.9/configparser.py", line 1254, in __getitem__
    raise KeyError(key)
KeyError: 'mt_gene_list'
```
The program reports an error because the keys corresponding to mt_gene_list and gtf cannot be found in the config file, not whether these two files exist. The correct celescope_genome.config should look like this
```
[files]
mt_gene_list = 'mt_gene_list.txt'
gtf = 'some gtf'
...
```
Can you show the contents of the celescope_genome.config file you generated?

**Reply by daniquebax** [reply]
Thanks for the quick reply!

The celescope_genome.config does indeed not include that data:

[files]
fasta = EBV_genome.fasta

[meta]
genome_name = EBV
genome_type = virus
celescope_version = 2.0.7
STAR_param =

**Reply by daniquebax** [reply]
I tried it again from scratch and now it works! :)

**Reply by zhouyiqi91** [reply]
Good to know. The previous `celescope_genome.config` corresponds to the EBV virus genome. By default there is no mt_gene_list and gtf.

**Reply by daniquebax** [reply]
Hi!

I hope you don't mind, but I have another question. I tried to run the pipeline now with my own data. I sequenced my DynaSCOPE libraries wich looked good on bioanalyzer and concentration. Multiple samples were pooled on the sequencing flowcell. So afterwards I demultiplexed the data with the adater sequences from the DynaSCOPE manual (I used cellranger which is uses on the bcl2fastq method from illumina).
However, I when I try to run the dynaseq pipeline I get an error and I think the main problem is that it says that my R1 is too short: GNGGATTGACGTTGATGTTGGTCTCTGAGTAGTCGACTTCGTATTAGCTTTGTAGGGCT. But its around 55 bp which is expected from the 2x 50 bp illumina kit?
Do you know a way to make this work?
[error DynaSCOPE.txt](https://github.com/singleron-RD/CeleScope/files/14485850/error.DynaSCOPE.txt)

**Reply by zhouyiqi91** [reply]
DynaSCOPE requires at least 72bp in R1 read to work. Below is the library structure:
`5' - read1(adapter) - bc1 - linker1 - bc2 - linker2 - bc3 - C - UMI - polyT - insert - TSO - 3'`
bc1=bc2=bc3=9bp
linker1=linker2=16bp
UMI=12bp
9 * 3+16 * 2+12+1 = 72

**Comment by Xiaomeng-011** [comment]
Hi Yiqi,
I tried to use your celescope rna mkref code to generate the ref files for several times, but I never got the SA and SA index file finally. The files I got are:
![WechatIMG88](https://github.com/user-attachments/assets/b8310392-0c81-4e1b-8fd3-5b035e680651)
and here is the log file: 
![image](https://github.com/user-attachments/assets/4560138d-4dce-48ac-8032-74bc85e8d607)
Coule you help me find out how can I solve this problem?
Thanks for your help.

**Reply by zhouyiqi91** [reply]
Building index may take a long time.

![image](https://github.com/user-attachments/assets/c8a9e108-d81e-4b81-a7ee-26afe9369c66)

**Comment by ianfmcn** [comment]
Hello! I have a question about the DynaSeq pipeline. In the conversion step, does the {sample}.PosTag.bam file only contain labeled reads, or does it label the reads in a way that makes it possible to separate out the labeled and unlabeled reads?

**Reply by zhouyiqi91** [reply]
@pigraul

**Reply by pigraul** [reply]
Hello! The `{sample}.PosTag.bam` file contains both labeled and unlabeled reads. They can be separated by BAM tags.

To distinguish between the two, you can look at the **TL** or **AL** tags in the BAM file.

* **TL tag**: Records the T to C mutation sites for a read. (eg. `TL:B:I,4259022,4259031`)
* **AL tag**: Records the A to G mutation sites.
* **ST tag**: Records the strand of the gene.

If the **ST** tag is `+`, you should look at the **TL** tag to find the T to C mutation sites. If the **ST** tag is `-`, you should look at the **AL** tag for the A to G mutation sites.

If a read has no mutation information, its TL or AL tag will be marked as `TL:B:C,0` / `AL:B:C,0` . You can use this information to identify the unlabeled reads.

**Comment by Gudaobeiguanren** [comment]
Hi,
I would appreciate it if you could tell me the difference between these two types of chemistry.
Chemistry：GEXSCOPE-V2  、Chemistry：flv_rna

**Reply by zhouyiqi91** [reply]
`GEXSCOPE-V2` is a chemistry used for standalone single-cell RNA experiments.  
`flv_rna` is a chemistry for single-cell RNA that is compatible with full-length VDJ.  
You can find the barcode and UMI structure here: https://github.com/singleron-RD/CeleScope/blob/master/doc/chemistry.md

**Reply by Gudaobeiguanren** [reply]
Thank you very much for your response. May I confirm whether my understanding is correct as follows?
GEXSCOPE-V2 is for transcriptome-only single-cell RNA-seq, whereas flv_rna allows the same batch of cells to be used for both transcriptome profiling and full-length VDJ (TCR/BCR) sequencing.

**Reply by Gudaobeiguanren** [reply]
flv_rna: In other words, the same batch of cells can be used for both transcriptome profiling and full-length V(D)J sequencing; the two share the same set of cell barcodes, allowing TCR/BCR clonotypes to be mapped back to individual cells.

**Reply by zhouyiqi91** [reply]
Exactly: the same batch of cells can be used for both transcriptome profiling and full-length V(D)J sequencing; the two share the same set of cell barcodes, allowing TCR/BCR clonotypes to be mapped back to individual cells.

**Reply by Gudaobeiguanren** [reply]
I see. Thank you very much.

**Comment by zhouyiqi91** [comment]
yes， you are right

----------

该邮件从移动设备发送









--------------原始邮件--------------
发件人："Gudaobeiguanren ***@***.***&gt;;
发送时间：2025年9月13日(星期六) 下午4:50
收件人："singleron-RD/CeleScope" ***@***.***&gt;;
抄送："周义其 ***@***.***&gt;;"Author ***@***.***&gt;;
主题：Re: [singleron-RD/CeleScope] Welcome to CeleScope Discussions! (Discussion #27)
-----------------------------------

 

 
Thank you very much for your response. May I confirm whether my understanding is correct as follows?
 GEXSCOPE-V2 is for transcriptome-only single-cell RNA-seq, whereas flv_rna allows the same batch of cells to be used for both transcriptome profiling and full-length VDJ (TCR/BCR) sequencing.

**Comment by matteobroketa** [comment]
Hi, I’m planning to use CeleScope for intentionally mixed human–mouse wells, including true mixed-species doublet barcodes, and I want to preserve their UMI signal as completely as possible.

I noticed the default EmptyDrops cell calling, `outFilterMatchNmin=50`, and STARsolo’s inherited defaults for one-mismatch UMI collapsing and unique-gene counting. However, I wonder if there are any other internal thresholds or assumptions, particularly around combined-species references, multimapping, UMI correction, or downstream filtering, that could drop or merge UMIs from these events? Would you recommend using the raw matrix or any particular parameter settings for this use case?

Thanks!

**Reply by zhouyiqi91** [reply]
Hi, please refer to https://github.com/singleron-RD/human_mouse

**Reply by matteobroketa** [reply]
Thanks!
Have you ever tested altering  --STAR_param "--soloMultiMappers ....." to run multiple methods (Uniform PropUnique EM Rescue) and generate multiple .mtx files? Or do you recommend individual runs?

**Reply by zhouyiqi91** [reply]
I have not tested different --soloMultiMappers options for human–mouse mixed samples. We usually rely on uniquely assigned reads for species classification, since multi-mapping reads generally have limited impact on human/mouse separation.
